### ISSUE TITLE
CFNTypes vs native Python types

### ISSUE THREAD

**Issue by mchristen** [remind101/stacker#339]
I noticed that in stacker_blueprints a lot of the base blueprints had their variable's types swapped out from types that inherit from CFNType to native python types and the code was updated to not use Ref and instead reach into the variables dict directly.

Is there a preferred way going forward? It was my understanding that the CFNTypes are newer and were going to be the preferred way of working with Variables vs Parameters and all of our internal blueprints use CFNTypes. However, the two different ways have nuanced behaviors that is not very apparent from the documentation.

```
1. Non CFNTypes aren't added to a template's Parameter section and can not be referenced
   via Ref('VariableName') and instead have to be looked up directly in the variables dict
2. list type is NOT compatible with CFNCommaDelimitedList.
    a. PublicSubnets: 10.128.1.0/24,10.128.2.0/24 <-- using CFNCommaDelimitedList
    b. PublicSubnets:    <-- using list
        - 10.128.0.0/24
        - 10.128.1.0/24
        - 10.128.2.0/24
        - 10.128.3.0/24
```

I'm basically at an impasse in upgrading to the latest stacker_blueprints because our blueprints rely heavily on CFNCommaDelimitedList but lots of the base blueprints we inherit from are now using native Python types.

**Comment by phobologic** [comment]
Hey @mchristen - we tend to lean towards using variables directly with the python types, though we've provided the CFNType's in case people still want to use Parameters (while allowing us to keep a single, consistent interface for defining both Variables & Parameter - since they're all Variables).

stacker_blueprints is meant to be an example repo that people could use, though you should expect changes as new stacker features roll out.  There's a few ways you could move forward with this:

1. Convert to using the variables with python types directly - we've done this ourselves internally, for example, with the RDS blueprints.  When doing anything that might cause changes, it's worth running with `--interactive` to verify nothing should change.  I've seen CFN Changesets think certain resources change, though nothing actually ends up changing with any of the physical resources.
2. You can "forward port" the old blueprint pretty easily to use variables.  Mostly you would change the `PARAMETERS` class attribute to `VARIABLES`, and then replace the `type` of each type with the appropriate CFNType.  That should make it a total no-op.

In general, I think going forward we will make most blueprints using python types in the variables - the one exception will likely be anything that has credentials in it, in which case we'll use a CFNType so that we can disable displaying of the value in the CFN console.

Hope that helps, and sorry for any confusion!

**Comment by mchristen** [comment]
Ok that makes more sense. I think my original assumption that CFNTypes were the way forward was incorrect. 

We already updated our blueprints to use VARIABLES over PARAMETERS awhile ago(I think prior to 1.0.0a5) but kept on using types based off of CFNType and had no issues until I went to upgrade to the latest released 1.0.0. 

Is there a technical reason for preferring the native types over CFNTypes? IMHO the blueprint code is easier to read when everything uses Ref and doesn't reach into the variables dict, there is less Python boilerplate and it is easier to leverage the Cloud Formation intrinsic functions.

**Comment by phobologic** [comment]
We've found native variables are easier to work with in a lot of cases - especially when you want to do things like create a variable # of things (for example, the VPC blueprint is so much easier than when we had to use Parameters - the proto-Variables idea came from LocalParameters, something we needed to make the vpc blueprint make sense - to make the # of AZs used dynamic).

In the end I think it's mostly a personal taste thing, and we'll continue to support the CFNTypes for the foreseeable future (I think it's a much nicer way of handling them than the way we did before).

I will say the one downside to direct variables is that it's harder to do the `diff` command.  That said, I plan to move `diff` to using change sets, much like the `--interactive` flag, since those are actually more complete (`diff` couldn't tell you if parameters changed very well, and honestly what we think a diff is, and what CFN thinks isn't always the same, and just trusting CFN is probalby the safest bet since it's kind of in charge).

I'm going to go ahead and close this out, but feel free to keep up the conversation here (or re-open it if you feel the need!).  Also, you should hop on the Slack team for stacker (you can see a badge to join it in the repo) :)
